Sources/HummingbirdJobs/JobQueue.swift

```diff
@@ -13,31 +13,79 @@
 //===----------------------------------------------------------------------===//
 
 import Foundation
-import Hummingbird
 import Logging
+import ServiceLifecycle
 
-/// Job queue protocol.
+/// Job queue
 ///
-/// Defines how to push and pop jobs off a queue
-public protocol HBJobQueue: AsyncSequence, Sendable where Element == HBQueuedJob<JobID> {
-    associatedtype JobID: CustomStringConvertible & Sendable
+/// Wrapper type to bring together a job queue implementation and a job queue
+/// handler. Before you can push jobs onto a queue you should register it
+/// with the queue via either ``HBJobQueue.registerJob(id:maxRetryCount:execute:)`` or
+/// ``HBJobQueue.registerJob(_:)``.
+public struct HBJobQueue<Queue: HBJobQueueDriver>: Service {
+    /// underlying driver for queue
+    public let queue: Queue
+    let handler: HBJobQueueHandler<Queue>
 
-    /// Called when JobQueueHandler is initialised with this queue
-    func onInit() async throws
-    /// Push Job onto queue
+    public init(_ queue: Queue, numWorkers: Int = 1, logger: Logger) {
+        self.queue = queue
+        self.handler = .init(queue: queue, numWorkers: numWorkers, logger: logger)
+    }
+
+    ///  Push Job onto queue
+    /// - Parameters:
+    ///   - id: Job identifier
+    ///   - parameters: parameters for the job
     /// - Returns: Identifier of queued job
-    @discardableResult func push(_ job: HBJob) async throws -> JobID
-    /// This is called to say job has finished processing and it can be deleted
-    func finished(jobId: JobID) async throws
-    /// This is called to say job has failed to run and should be put aside
-    func failed(jobId: JobID, error: any Error) async throws
-    /// stop serving jobs
-    func stop() async
-    /// shutdown queue
-    func shutdownGracefully() async
+    @discardableResult public func push<Parameters: Codable & Sendable>(id: HBJobIdentifier<Parameters>, parameters: Parameters) async throws -> Queue.JobID {
+        let jobRequest = HBJobRequest(id: id, parameters: parameters)
+        let data = try JSONEncoder().encode(jobRequest)
```

Comment:
Would it be possible to allow drivers to encode these parameters (or the request as a whole)? In MongoDB, it would be really unfortunate to encode this to JSON, then add the JSON into MongoDB. Since MongoDB already has (recursive) object support. So it would be able to store this as-is in BSON. Encoding to JSON would prevent indexes from being applies to `id`, and inspecting the data in MongoDB would require quite a lot of work from the developer.

Reply:
The idea is that the driver only stores data blobs. Do you see an advantage to storing the full object structure? I guess debugging job issues might be easier.

Reply:
Debugging mainly, yes

Reply:
I had moved away from the allowing drivers to do the encoding because I was repeating the same code across all of the drivers. It also meant the driver needed to know about the job registry so they could be sure to decode the correct job type. Data blobs meant the driver had a clean split from the job decoding. I'll have another look though.